url5583f3dbb2_XOR19_XScript_tgz-pJ8-xscript_interactive_MainJ8/src/xscript/compiler/XTreeCompiler.java

```diff
@@ -0,0 +1,1298 @@
+		part.setup(this, code);
+		part.set(this, code);
+		visitInScope(foreach.body).asStatement(this, code);
+		code.addInstruction(foreach, XOpcode.JUMP, _continue);
+		code.addInstructionB(foreach, XOpcode.POP, this.scope.getLocalsCount());
+		scope = scope.getParent();
+		code.addInstruction(foreach, end);
+		code.addInstruction(foreach, XOpcode.POP_1);
+		code.addInstruction(foreach, _break);
+	}
+
+	@Override
+	public void visitWhile(XTreeWhile _while) {
+		XCodeGen code = getCode(_while);
+		XJumpTarget _break = new XJumpTarget();
+		XJumpTarget _continue = new XJumpTarget();
+		scope = new XScope(scope, true, _break, _continue, labels, 0);
+		labels = null;
+		code.addInstruction(_while, _continue);
+		XCompiledPart part = visit(_while.condition);
+		if (part.isConstValue()) {
+			if (part.isFalse()) {
+				addDiagnostic(_while.body, "dead.code");
+			} else {
+				visitInScope(_while.body).asStatement(this, code);
+				code.addInstruction(_while, XOpcode.JUMP, _continue);
+				if (_break.jumps() == 0)
+					treePart.setDead(true);
+			}
+		} else {
+			part.setupAndGet(this, code);
+			code.addInstruction(_while, XOpcode.JUMP_IF_ZERO, _break);
+			visitInScope(_while.body).asStatement(this, code);
+			code.addInstruction(_while, XOpcode.JUMP, _continue);
+		}
+		code.addInstructionB(_while, XOpcode.POP, this.scope.getLocalsCount());
+		scope = scope.getParent();
+		code.addInstruction(_while, _break);
+	}
+
+	@Override
+	public void visitDo(XTreeDo _do) {
+		XCodeGen code = getCode(_do);
+		XJumpTarget _break = new XJumpTarget();
+		XJumpTarget _continue = new XJumpTarget();
+		XJumpTarget start = new XJumpTarget();
+		scope = new XScope(scope, true, _break, _continue, labels, 0);
+		labels = null;
+		code.addInstruction(_do, start);
+		visitInScope(_do.body).asStatement(this, code);
+		code.addInstruction(_do, _continue);
+		XCompiledPart part = visit(_do.condition);
+		if (part.isConstValue()) {
+			if (!part.isFalse()) {
+				code.addInstruction(_do, XOpcode.JUMP, start);
+				if (_break.jumps() == 0)
+					treePart.setDead(true);
+			}
+		} else {
+			part.setupAndGet(this, code);
+			code.addInstruction(_do, XOpcode.JUMP_IF_NON_ZERO, start);
+		}
+		code.addInstructionB(_do, XOpcode.POP, this.scope.getLocalsCount());
+		scope = scope.getParent();
+		code.addInstruction(_do, _break);
+	}
+
+	@Override
+	public void visitSynchronized(XTreeSynchronized _synchronized) {
+		XCodeGen code = getCode(_synchronized);
+		visit(_synchronized.sync).asStatement(this, code);
+		visit(_synchronized.body).asStatement(this, code);
+		addDiagnostic(_synchronized, "synchronized.not.implemented");
+	}
+
+	@Override
+	public void visitBreak(XTreeBreak _break) {
+		XCodeGen code = getCode(_break);
+		XJump jump = new XJump();
+		if (!scope.getJump(_break.lable == null ? null : _break.lable.name, 0, jump)) {
+			addDiagnostic(_break, "break.not.found");
+		} else {
+			jump.gen(_break, code);
+		}
+		treePart.setDead(true);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 491:498, 503:510.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.